WooCommerce/Classes/ViewRelated/Orders/Order Details/Order Notes Section/Customer Note/EditCustomerNoteViewModel.swift

```diff
@@ -84,14 +91,111 @@ final class EditCustomerNoteViewModel: EditCustomerNoteViewModelProtocol {
 private extension EditCustomerNoteViewModel {
     /// Calculates what navigation trailing item should be shown depending on our internal state.
     ///
-    private func bindNavigationTrailingItemPublisher() {
+    func bindNavigationTrailingItemPublisher() {
         Publishers.CombineLatest($newNote, performingNetworkRequest)
-            .map { [order] newNote, performingNetworkRequest -> EditCustomerNoteNavigationItem in
-                guard !performingNetworkRequest else {
+            .map { [weak self] newNote, performingNetworkRequest -> EditCustomerNoteNavigationItem in
+                let optimisticUpdatesEnabled = self?.areOptimisticUpdatesEnabled ?? false
+                guard optimisticUpdatesEnabled || !performingNetworkRequest else {
                     return .loading
                 }
-                return .done(enabled: order.customerNote != newNote)
+                return .done(enabled: self?.order.customerNote != newNote)
             }
             .assign(to: &$navigationTrailingItem)
     }
+
+    /// Indicates whether the optimistic updates are enabled.
+    ///
+    var areOptimisticUpdatesEnabled: Bool {
+        featureFlagService.isFeatureFlagEnabled(.updateOrderOptimistically)
+    }
+
+    /// Updates the temporal note after updating the order.
+    ///
+    func syncNewNoteWithOrder() {
+        newNote = order.customerNote ?? ""
+    }
+
+    /// Handles the action to update the order.
+    /// - Parameters:
+    ///   - customerNote: Given new customer note to update the order.
+    ///   - onFinish: Callback to notify when the action has finished.
+    ///
+    func handleOrderUpdate(withNote customerNote: String?, onFinish: ((Bool) -> Void)? = nil) {
+        let orderID = order.orderID
+        let modifiedOrder = order.copy(customerNote: customerNote)
+
+        let updateAction = makeUpdateAction(order: modifiedOrder) { [weak self] result in
+            guard let self = self else {
+                onFinish?(result.isSuccess)
```

Comment:
I wonder if it's necessary to trigger the completion closure here, we usually just return early when this happens. WDYT?

Reply:
mmm...good question 🤔 At first I thought this might be a good idea in case there is a scenario where the view is still alive even if the view model is not. But I think it's not possible because if `EditCustomerNoteViewModel` doesn't exist is because the Order details view model doesn't exist either. That means, the Order details screen doesn't exist and  it wouldn't make sense to display this feedback.

That said, I agree with you that we could just return early. Let me change it 👍🏼

Thanks! 🙂